[GH-17861] Modify the playbook run list status filter to be a multiselect like the RHS timeline filter

nika-begiashvili opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository-specific documentation at https://developers.mattermost.com/contribute/getting-started/

REMEMBER TO:
- Run `make check-style` to check for style errors (required for all pull requests)
- Run `make test` to ensure unit tests passed
-->

#### Summary
<!--
A description of what this pull request does
-->

This PR adds capabilities to filter playbook run list using multiple statuses

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.:

  Fixes: https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the Jira ticket.
-->

https://github.com/mattermost/mattermost-server/issues/17861

#### Checklist
<!-- Check off items as they are completed. ~~Strike through~~ items if they don't apply -->
- [ ] Telemetry updated
- [ ] Gated by experimental feature flag
- [x] Unit tests updated

mattermod commented:
Hello @nika-begiashvili,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

iomodo reviewed (commented):
Great work @nika-begiashvili! Thanks for the contribution!
Couple of questions below.

Also I noticed that by default checkboxes are checked. Might it be better to uncheck all checkboxes and by default fetch all runs? @abhijit-singh @itao we will need your input here.

cpoile commented on webapp/src/client.ts:
```diff
@@ -42,7 +42,7 @@ import {GlobalSettings, globalSettingsSetDefaults} from './types/settings';
 
 export async function fetchPlaybookRuns(params: FetchPlaybookRunsParams) {
-    const queryParams = qs.stringify(params, {addQueryPrefix: true});
+    const queryParams = qs.stringify(params, {addQueryPrefix: true, indices: false});
```
so that an array of query params are not postfixed with indexes. Go can't parse params with indexes, but it will construct a slice when multiple params have the same key.

nika-begiashvili commented on webapp/src/components/backstage/playbook_runs/playbook_run_list/status_filter.tsx:
```diff
@@ -20,118 +18,55 @@ export interface StatusOption {
+                newOptions.filter((opt) => opt.selected && opt.value !== '').map((opt) => opt.value),
+            );
+        } else {
+            props.onChange(['_']); // no status is selected, should return empty list
```
currently I have 3 general cases in regards to sqlstore 

- get all statuses, in this case we shouldn't add 'where' clause at all - frontend sends empty array in this case
- get one or more statuses, we add 'where' clause - frontend sends corresponding statuses as array
- none of the statuses are selected, should query non existent status or not query at all - frontend sends ['_'] (non-existent status)

I understand that third case can be handled other ways, I'm open for suggestions

nika-begiashvili commented on server/api/playbook_runs.go:
```diff
@@ -1322,7 +1322,8 @@ func parsePlaybookRunsFilterOptions(u *url.URL) (*app.PlaybookRunFilterOptions,
 	direction := u.Query().Get("direction")
 
-	status := u.Query().Get("status")
+	//status := u.Query().Get("status")
```
done

itao commented:
> Also I noticed that by default checkboxes are checked. Might it be better to uncheck all checkboxes and by default fetch all runs? @abhijit-singh @itao we will need your input here.

I suggest default state being unchecking Archived but leave everything else checked so that it shows what is currently in progress.

cpoile reviewed (dismissed):
Really nice @nika-begiashvili !

As we discussed in our chat:
- clicking 'All' should grey out the other options that were selected (and not select any new options)
- un-clicking 'All' should unclick 'All' and un-grey the other options (so that it remembers the previous filter state)
- it would be great to add a divider between `All` and the rest of the options.

(if you have time)
- can we add telemetry so that we know if people are using the statuses multiselect?

nika-begiashvili commented:
> 
> 
> > Also I noticed that by default checkboxes are checked. Might it be better to uncheck all checkboxes and by default fetch all runs? @abhijit-singh @itao we will need your input here.
> 
> I suggest default state being unchecking Archived but leave everything else checked so that it shows what is currently in progress.

Thank you! I made changes according to your feedback

nika-begiashvili commented on webapp/src/components/backstage/playbook_runs/playbook_run_list/status_filter.tsx:
```diff
@@ -3,14 +3,12 @@
-import ReactSelect, {StylesConfig} from 'react-select';
-import styled from 'styled-components';
-
+import {MultiSelect, SelectOption} from '../../../multi_select';
```
fixed relative path

nika-begiashvili commented:
Hi, I've made requested changes: modified 'All' checkbox user experience and added divider
but it seems one e2e test is failing which doesn't look like is related to my changes, can anyone assist me with that ?

lieut-data commented:
@nika-begiashvili, there's a non-zero chance it was a flaky test. I've kicked the step to see if it passes on second run.

crspeller reviewed (approved):
Looks good! The e2e issues is happening with other PRs, so it's not you :)

iomodo reviewed (approved):
Looks great @nika-begiashvili! Thanks!

lieut-data commented:
@crspeller, is it ok to dismiss @cpoile's review as he's OOO?

abhijit-singh commented:
Thanks for tackling this @nika-begiashvili! 

Will it be possible to attach a few screenshots/recording here to show how the filter is working?

nika-begiashvili commented:
> 
> 
> Thanks for tackling this @nika-begiashvili!
> 
> Will it be possible to attach a few screenshots/recording here to show how the filter is working?

Sure, here are some screenshots

![image](https://user-images.githubusercontent.com/259598/125574355-dc67e798-f96a-4d36-b4c1-a0844bf0e8ca.png)
![image](https://user-images.githubusercontent.com/259598/125574389-b77cc5d7-a7ac-4bad-b2a6-98357ae64fcf.png)
![image](https://user-images.githubusercontent.com/259598/125574461-7c14ac2f-b239-4aa2-b826-5eb7f67bbc7b.png)

abhijit-singh commented:
@nika-begiashvili Thank you! A couple of small points:

1. We currently have a `hover` state and an `active` for the filter buttons. The active state seems to be missing in the screenshots. Can we incorporate both of those states? 
![2021-07-14 12 21 30](https://user-images.githubusercontent.com/14369607/125577136-b4fc7f03-7a14-4abd-a4ef-69ed8a0f486c.gif)

2. Just wanted to make sure that in the dropdown menu clicking anywhere on a row should check/uncheck the respective item.

nika-begiashvili commented:
@abhijit-singh apologies, screenshots were not very useful, here's video I've captured


https://user-images.githubusercontent.com/259598/125582122-4f3ab2b8-f369-457e-9852-b1217f9a7887.mp4

abhijit-singh commented:
@nika-begiashvili This looks good! One last point, can we make the `active` state for the button persist even when the dropdown is open? Just like we have today.
![image](https://user-images.githubusercontent.com/14369607/125611639-fb1c7ef8-3651-4351-bf4f-d2575d513026.png)

nika-begiashvili commented:
@abhijit-singh like this ?


https://user-images.githubusercontent.com/259598/125669101-270a8419-e9ab-49ef-b000-73c5f09d6c1f.mp4

abhijit-singh reviewed (approved):
Looks good! Thank you @nika-begiashvili!
